### DISCUSSION TITLE
Fore 3.0

### DISCUSSION THREAD

**Discussion by JoernT** [Jinntec/Fore#272] [General]
## a glimpse towards Fore 3.0

Maybe there will another 2.x release but i started some work towards 3.0. 

This release will be breaking by removing some `<fx-` elements where a better, native HTML alternative exists. For instance we've discussed the `<fx-instance>` element before as being unluckily named - it always requires extra explanation to say `it's a data-container`. Further HTML already defines the `<data>` element. Though it's purpose is to define key/value pairs mainly it does not prohibit child content and would be a more natural match for the purpose. This means that `<fx-instance>` will just be replaced with the native HTML element.

Example:
```xml
<fx-fore>
    <fx-model>
         <!-- instead of ... -->
         <fx-instance id="mydata" src="mysource.xml"></fx-instance>
         <!-- use data element instead -->
         <data id="mydata" data-src="mysource.xml"></data>
    </fx-model>
...
</fx-fore>
```

This comes to the price to use `data-` prefixes for the attributes (as the standard requires for custom attributes) but otherwise is more expressive than the old syntax and builds directly on an already existing element. 

This is probably only one change that will go into 3.0. The `<fx-dialog>` is another candidate as HTML defines this also in the meantime. - we'll critically review other candidates. The rationale behind it is to give already existing elements priority over custom elements and simplify either the codebase and the usage of Fore by building upon the common ground.

### drop it or keep it

Of course the question arises if we should just keep the `<fx-instance>` element and add the use of `<data>` as an additional option. However i decided against it:

* it takes much more effort to keep them both as it requires some code-branching and duplicate functionality 
* it would encourage a mixed use of both elements not doing clarity and maintainability a favor
* it would also lead to a mix in the demos likely leading to confusion for users
* the documentation would also take this into account
* just deprecating now and removing later does only push the problem into the future and cost us precious dev-time not being spend elsewhere

Enough reasons i hope to justify a 'hard' break requiring users to update their pages when upgrading to 3.0

Making these changes is still a pile of work but at least will shrink the codebase and bring us closer to the platform.

Unfortunately the comments are not migrated from the original a discussion - please see the link below if interested.
_Originally posted by @JoernT in https://github.com/Jinntec/Fore/discussions/269_

## Conclusions from discussions so far

* The `<fx-instance>` element is dropped and replaced with HTML `<data>` element
* Adressing data will be done by variables instead of a function as before with `instance()` function. For each data element there will be a predefined variable using the id as its name. 
* the default data are reachable by $default

**Comment by JoernT** [comment]
Another one mentioned by Juri - might make sense to keep the instance() function as before but resolving to the variable instead - this would keep the migration effort lower as it's not required to change all over the UI at once to get it running with 3.0.

Interesting aspects at least ... maybe also a migration script would be helpful.

**Comment by rkottmann** [comment]
I am still not that much into fore being able to add value to this discussion.

However, I love the spririt:

> Making these changes is still a pile of work but at least will shrink the codebase and bring us closer to the platform.

**Reply by JoernT** [reply]
To be honest this effort has stalled at the moment in favor of other improvements we're doing mainly for performance. Despite these changes looking innocent they will be a large work package to do - just to migrate all the demos beyond the rework of the code. I fear efforts in this direction will need to wait for a project at least partly paying the cost.
